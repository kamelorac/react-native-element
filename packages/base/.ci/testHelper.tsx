import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
