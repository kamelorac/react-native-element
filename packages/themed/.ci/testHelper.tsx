import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import {
  ThemeProvider,
  FullTheme,
  createTheme,
  CreateThemeOptions,
} from '../src/config';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<CreateThemeOptions> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    <ThemeProvider theme={createTheme(themeProp)}>{children}</ThemeProvider>,
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
