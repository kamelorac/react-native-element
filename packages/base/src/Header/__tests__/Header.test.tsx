import React from 'react';
import { Header } from '../index';
import { Button, ImageBackground, StatusBar, View } from 'react-native';
import { renderWithWrapper } from '../../../.ci/testHelper';

const btnCfg = { icon: 'home' };
const titleCfg = { text: 'This is a title' };

describe('Header Component', () => {
  it('should match snapshot', () => {
    const { toJSON } = renderWithWrapper(
      <Header
        backgroundImage={{
          uri: 'https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885_1280.jpg',
        }}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render children when passed in', () => {
    const { wrapper } = renderWithWrapper(
      <Header>
        <Button title="Test button" onPress={() => {}} />
      </Header>
    );
    expect(wrapper.findAllByType(Button).length).toBe(1);
    expect(wrapper.findByType(Button)).not.toBeNull();
  });

  it('should render multiple children when passed in', () => {
    const { wrapper } = renderWithWrapper(
      <Header>
        <Button title="Test button 1" onPress={() => {}} />
        <Button title="Test button 2" onPress={() => {}} />
      </Header>
    );
    expect(wrapper.findAllByType(Button).length).toBe(2);
  });

  it('should render left component by passing a config through props', () => {
    const component = renderWithWrapper(<Header leftComponent={btnCfg} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render left component by passing a component through props', () => {
    const component = renderWithWrapper(
      <Header
        leftComponent={<Button title="Test button" onPress={() => {}} />}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render right component by passing a config through props', () => {
    const component = renderWithWrapper(<Header rightComponent={btnCfg} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render right component by passing a component through props', () => {
    const { wrapper } = renderWithWrapper(
      <Header
        rightComponent={<Button title="Test button" onPress={() => {}} />}
      />
    );
    expect(wrapper.findByType(Button).props.title).toBe('Test button');
  });

  it('should render center component by passing a config through props', () => {
    const component = renderWithWrapper(<Header centerComponent={titleCfg} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render center component by passing a component through props', () => {
    const component = renderWithWrapper(
      <Header
        centerComponent={<Button title="Test button" onPress={() => {}} />}
      />
    );
    expect(component.wrapper.findByType(Button).props.title).toBe(
      'Test button'
    );
  });

  it('should allow to pass backgroundColor through prop', () => {
    const { wrapper } = renderWithWrapper(<Header backgroundColor="#aaa" />);
    expect(wrapper.findAllByType(View)[0].props.style.backgroundColor).toBe(
      '#aaa'
    );
  });

  it('should have valid elevate prop', () => {
    const { wrapper } = renderWithWrapper(<Header elevated />);
    expect(wrapper.findAllByType(View)[0].props.style.elevation).toBe(24);
  });

  it('should allow to pass styles through containerStyle prop', () => {
    const { wrapper } = renderWithWrapper(
      <Header containerStyle={{ backgroundColor: '#ccc' }} />
    );
    expect(wrapper.findAllByType(View)[0].props.style.backgroundColor).toBe(
      '#ccc'
    );
  });

  it('should accept props for StatusBar', () => {
    const { wrapper } = renderWithWrapper(
      <Header statusBarProps={{ hidden: true }} />
    );
    expect(wrapper.findByType(StatusBar).props.hidden).toBe(true);
  });

  it('should allow to pass backgroundImageSource through prop', () => {
    const { wrapper } = renderWithWrapper(
      <Header backgroundImage={{ uri: 'http://google.com' }} />
    );
    expect(wrapper.findByType(ImageBackground).props.source).toEqual({
      uri: 'http://google.com',
    });
  });

  it('should allow to pass backgroundImageStyle through prop', () => {
    const { wrapper } = renderWithWrapper(
      <Header
        backgroundImage={{ uri: 'http://google.com' }}
        backgroundImageStyle={{ opacity: 0.1 }}
      />
    );
    expect(wrapper.findByType(ImageBackground).props.imageStyle).toMatchObject({
      opacity: 0.1,
    });
  });
});
