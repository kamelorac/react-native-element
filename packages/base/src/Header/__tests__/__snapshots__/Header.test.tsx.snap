// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2089dc",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://cdn.pixabay.com/photo/2015/04/23/22/00/tree-736885_1280.jpg",
        }
      }
      style={
        [
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          {
            "height": undefined,
            "width": undefined,
          },
          undefined,
        ]
      }
      testID="headerContainer"
    />
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      />
    </RNCSafeAreaView>
  </View>
</View>
`;

exports[`Header Component should render center component by passing a config through props 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2089dc",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="headerContainer"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      >
        <Text
          numberOfLines={1}
          text="This is a title"
        >
          This is a title
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      />
    </RNCSafeAreaView>
  </View>
</View>
`;

exports[`Header Component should render left component by passing a component through props 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2089dc",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="headerContainer"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              [
                {},
              ]
            }
          >
            <Text
              style={
                [
                  {
                    "color": "#007AFF",
                    "fontSize": 18,
                    "margin": 8,
                    "textAlign": "center",
                  },
                ]
              }
            >
              Test button
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      />
    </RNCSafeAreaView>
  </View>
</View>
`;

exports[`Header Component should render left component by passing a config through props 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2089dc",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="headerContainer"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "flex-start",
              "overflow": "hidden",
            }
          }
          testID="RNE__ICON__CONTAINER"
        >
          <View
            accessibilityRole="button"
            android_ripple={
              {
                "borderless": false,
                "color": "rgba(0, 0, 0, 0.3)",
                "radius": -5,
              }
            }
            disabled={false}
            icon="home"
            testID="RNE__ICON__CONTAINER_ACTION"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
              testID="RNE__ICON"
            >
              <Text
                allowFontScaling={false}
                style={
                  [
                    {
                      "color": "#242424",
                      "fontSize": 24,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="RNE__ICON__Component"
              >
                î¢Š
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      />
    </RNCSafeAreaView>
  </View>
</View>
`;

exports[`Header Component should render right component by passing a config through props 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#2089dc",
        "borderBottomColor": "#f2f2f2",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 10,
        "paddingVertical": 10,
      }
    }
    testID="headerContainer"
  >
    <RNCSafeAreaView
      edges={
        [
          "left",
          "top",
          "right",
        ]
      }
      style={
        {
          "flexDirection": "row",
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flex": 1,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "center",
            "flex": 3,
          }
        }
      />
      <View
        style={
          {
            "alignItems": "flex-end",
            "flex": 1,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "flex-end",
              "overflow": "hidden",
            }
          }
          testID="RNE__ICON__CONTAINER"
        >
          <View
            accessibilityRole="button"
            android_ripple={
              {
                "borderless": false,
                "color": "rgba(0, 0, 0, 0.3)",
                "radius": -5,
              }
            }
            disabled={false}
            icon="home"
            testID="RNE__ICON__CONTAINER_ACTION"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "transparent",
                  "justifyContent": "center",
                }
              }
              testID="RNE__ICON"
            >
              <Text
                allowFontScaling={false}
                style={
                  [
                    {
                      "color": "#242424",
                      "fontSize": 24,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                    {
                      "fontFamily": "Material Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
                testID="RNE__ICON__Component"
              >
                î¢Š
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;
